faithful signature support in BoxedKernelWrapper

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47267 faithful signature support in BoxedKernelWrapper**

### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

Commit:
faithful signature support in BoxedKernelWrapper

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


Still needs single-result faithful support, don't bother reviewing yet.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention: for boxed kernels, the wrappers provide unboxed calling - taking their args directly, pushing them onto an IValue stack, calling the boxed kernel, popping the result(s) off the stack and returning them. This means that any op can be called using either convention, irrespective of how its kernel was implemented.

An ops falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but thee system expects to be able to make unboxed calls to them with out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front)
- legacy out-of-place multi (multiple tensor refs at the front)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`
* tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`

The rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention: for boxed kernels, the wrappers provide unboxed calling - taking their args directly, pushing them onto an IValue stack, calling the boxed kernel, popping the result(s) off the stack and returning them. This means that any op can be called using either convention, irrespective of how its kernel was implemented.

An ops falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention: for boxed kernels, the wrappers provide unboxed calling - taking their args directly, pushing them onto an IValue stack, calling the boxed kernel, popping the result(s) off the stack and returning them. This means that any op can be called using either convention, irrespective of how its kernel was implemented.

An ops falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -236,30 +231,76 @@ struct BoxedKernelWrapper<
-// 4. signatures returning a tuple of Tensor references, and taking the same
-// number of Tensor refs as their initial arguments.
+// 4. an out of place op takes one or more non-const Tensor ref arguments at the
+// end of the arg list, and also returns it/them. This specialization supports
```
Well, this specialization only supports a single return, no?

bhosmer replied:
> Well, this specialization only supports a single return, no?

yup - see the sentence beginning with "This specialization supports..." 😀 

Fair point though, starting the comment this way is confusing, I'll reword

ezyang commented:
OK, I think I understand the approach taken here, and agree that it works. That's probably good enough to put it in.  (@smessmer should get a chance to review though).

However, I have some longevity concerns about this code. In particular, it seems that we will be in a lot of trouble if we ever fix the kernel calling convention to stop passing mutable tensor arguments by mutable reference. Then the piece of static information you're using to key these specializations over will no longer work. What I'm kind of getting the feeling from this is that maybe we should be OK with adding extra static metadata at dispatcher call sites that let templating key over it (something like Tensor&, but not actually that). This metadata can be checked for consistency with schema, and will help you drive more complicated template work. This might also simplify some problems that smessmer has.

bhosmer commented:
> However, I have some longevity concerns about this code. 

Shared. Pattern matching like this is a recipe for ongoing expense. It'll get simpler once the legacy ordering no longer needs handling, but it's not a long term solution in any case. 

I'm not sure the solution is metadata at call sites vs at registration time, but some sort of explicit classifier in any event. (in/out/func is probably all that's needed to make this particular thing robust)

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -194,34 +196,27 @@ struct BoxedKernelWrapper<
 
 template <class... OtherArgs>
 struct BoxedKernelWrapper<
   at::Tensor&(at::Tensor&, OtherArgs...),
```
that means this specialization only catches single-out-arg ops, not inplace ops. You might also have to rename test cases that you currently call "inplace".

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -236,30 +231,75 @@ struct BoxedKernelWrapper<
+  at::Tensor&(FirstArg, RestArgs...),
+  std::enable_if_t<
+    can_box_all<FirstArg, RestArgs...>::value
+    // this skips over in-place (and legacy out-of-place) kernels with a non-const Tensor
```
it might triggers for inplace

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -236,30 +231,75 @@ struct BoxedKernelWrapper<
+    can_box_all<Args...>::value && is_tuple_of_mutable_tensor_refs<Result>::value
+    // this test skips over legacy kernels with out args at the front, so they can trigger
+    // the specialization that follows.
+    // note: this test is complicated by the fact that boolean value expressions in templates
```
ugh...wishing for C++17 once again

smessmer reviewed (approved):
As said in chat, this does not reorder out arguments for non-c10-full ops correctly. It will calm the compiler for non-c10-full ops though, which is the most important thing for now. We can either ignore non-c10-full ops and wait until everything is c10-full, or we can try to fix the reordering in a separate PR.

Commit:
Update on "faithful signature support in BoxedKernelWrapper"


### Preliminaries: 

**Boxed/unboxed kernels vs. boxed/unboxed calling:**

Op kernels can be *boxed* or *unboxed*: the former are C++ functions that take an IValue stack containing the op arguments; the latter are C++ functions that take op arguments directly, as C++ parameters. 

In either case, the system creates wrappers that support the "other" calling convention, so that any op can be called using either convention, irrespective of how its kernel was implemented. 

So for a boxed kernel, the system creates an unboxed wrapper - a C++ function that takes the op args directly, pushes them onto an IValue stack, calls the boxed kernel, pops the result(s) off the stack and returns it/them.

**Schema varieties:**

An op falls into one of three categories, depending on its schema: 
- functional: args are immutable and return values are freshly created
- in place: a 'self' argument is passed by reference at the *start* of the argument list, mutated by the op, and returned by reference
- out of place: one or more "out" arguments are passed by reference at the *end* of the argument list. The op writes computed values to them and returns them by reference (in a tuple, if there's more than one).

To complicate things, there is an additional "legacy out of place" variation. These ops define out of place schemas per above, but the dispatcher expects their kernels to accept out arguments at the *start* of the list. (History behind this is out of scope for this description, but this category will eventually vanish.)

### This PR:

Here we extend support for unboxed wrappers around boxed kernels to include all in-place and out-of-place variations described above:

- in-place (single tensor ref at the front)
- out-of-place single (single tensor ref at the back)
- out-of-place multi (multiple tensor refs at the back, returned in a tuple)
- legacy out-of-place single (single tensor ref at the front, same as in-place)
- legacy out-of-place multi (multiple tensor refs at the front, returned in a tuple)

New tests added to cover these cases.

How to read:

* changes that matter are in `aten/src/ATen/core/boxing/impl/boxing.h`: new specializations of `BoxedKernelWrapper` cover (non-legacy) out of place signatures, previously unhandled
* new tests are in `aten/src/ATen/core/boxing/KernelFunction_test.cpp`, plus some renaming for clarity
* the rest is minor - minor helpers and tests, slight jitter in client code due to opportunistic rationalization in boxing.h.

Differential Revision: [D24701488](https://our.internmc.facebook.com/intern/diff/D24701488)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -194,34 +196,27 @@ struct BoxedKernelWrapper<
 //
-// 3. signatures taking a single Tensor reference as their first argument,
-// and also returning one.
+// 3. in-place and legacy out-of-place ops take a single non-const Tensor
```
For posterity - turns out in-place ops do in fact use `Tensor& self`. This makes this case ambiguous w.r.t. reordering when pushing onto the stack (in-place shouldn't, out-of-place should). Which means the commentary here is accurate, but leaves open the problem mentioned in Sebastian's comment. Looking at options but meanwhile cc @bdhirsh, this means you're still blocked on non-c10-full out ops, unfortunately. We can go over the situation at the next standup.
